### ISSUE TITLE
Crash when using Tasty Reflect extension methods and a QuoteContext from a class

### ISSUE THREAD

**Issue by anatoliykmetyuk** [lampepfl/dotty#8623]
# Example 1
```Scala
import scala.quoted._

class Dummy(val ctx: QuoteContext)
given (using h: Dummy) as QuoteContext = h.ctx

def f[T](expr: Expr[T])(using h: Dummy) = {
  import h.ctx.tasty._
  expr.unseal.pos.sourceCode
}
```

<details><summary>Crash output</summary>

```scala
exception occurred while compiling /Users/kmetiuk/Projects/scala3/pg/misc/iss_1.scala
java.lang.AssertionError: assertion failed: unresolved symbols: value <skolem>(line -1), value <skolem>(line -1), value <skolem>(line -1) when pickling /Users/kmetiuk/Projects/scala3/pg/misc/iss_1.scala while compiling /Users/kmetiuk/Projects/scala3/pg/misc/iss_1.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: unresolved symbols: value <skolem>(line -1), value <skolem>(line -1), value <skolem>(line -1) when pickling /Users/kmetiuk/Projects/scala3/pg/misc/iss_1.scala
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:722)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$9$$anonfun$8(Pickler.scala:63)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:305)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:87)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:305)
	at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:87)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:318)
	at scala.collection.immutable.List.map(List.scala:219)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:319)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:91)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:167)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:177)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:185)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:192)
	at dotty.tools.dotc.Run.compileSources(Run.scala:129)
	at dotty.tools.dotc.Run.compile(Run.scala:112)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:36)
	at dotty.tools.dotc.Driver.process(Driver.scala:189)
	at dotty.tools.dotc.Driver.process(Driver.scala:158)
	at dotty.tools.dotc.Driver.process(Driver.scala:170)
	at dotty.tools.dotc.Driver.main(Driver.scala:197)
	at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

Typer phase output:

```scala
    def f[T >: Nothing <: Any](expr: quoted.Expr[T])(using h: Dummy): String =
      {
        import h.ctx.tasty._
        <skolem>.tasty.positionOps.sourceCode(
          <skolem>.tasty.TreeOps.pos(
            expr.unseal(iss_1$package.given_QuoteContext(h))
          )(<skolem>.tasty.given_Context)
        )
      }
```

# Example 2
```scala
import scala.quoted._

class Dummy(val ctx: QuoteContext)
given (using h: Dummy) as QuoteContext = h.ctx

def f[T](expr: Expr[T])(using h: Dummy) =
  import h.ctx.tasty._
  expr.unseal.pos
```

Compile output:

```scala
-- [E007] Type Mismatch Error: /Users/kmetiuk/Projects/scala3/pg/misc/iss_1-1.scala:7:2
7 |  import h.ctx.tasty._
  |  ^
  |  Found:    ?1.tasty.Position
  |  Required: quoted.QuoteContext#tasty.Position
  |
  |  where:    ?1 is an unknown value of type quoted.QuoteContext
8 |  expr.unseal.pos
```

Typer phase output:

```scala
    def f[T >: Nothing <: Any](expr: quoted.Expr[T])(using h: Dummy):
      quoted.QuoteContext#tasty.Position
     =
      {
        import h.ctx.tasty._
        <skolem>.tasty.TreeOps.pos(
          expr.unseal(iss_1-1$package.given_QuoteContext(h))
        )(<skolem>.tasty.given_Context)
      }
```

**Comment by nicolasstucki** [comment]
Side note. We should never keep a `QuoteContext` in a field. It should always be passed as a parameter. Otherwise, it risks losing its contextual nature.

**Comment by nicolasstucki** [comment]
Self-contained example:

```scala
class QC:
  object tasty {
    type Tree
    type Term <: Tree
    type Context
    type Position
    given Context = ???
    extension TreeOps on (tree: Tree) {
      /** Position in the source code */
      def pos(using ctx: Context): Position = ???
    }
  }

class E[+T]:
  def unseal(using qctx: QC): qctx.tasty.Term = ???

class Dummy(val ctx: QC)
given (using h: Dummy) as QC = h.ctx

def f[T](expr: E[T])(using h: Dummy) = {
  import h.ctx.tasty._
  expr.unseal.pos
}
```

**Comment by nicolasstucki** [comment]
Workaround
```scala
given (using h: Dummy) as h.ctx.type = h.ctx
```

**Comment by nicolasstucki** [comment]
Minimized to
```scala
trait QC:
  object tasty:
    type Tree
    extension TreeOps on (tree: Tree):
      def pos: Tree = ???

def unseal(using qctx: QC): qctx.tasty.Tree = ???

given QC = ???

def test = unseal.pos
```

<details>

```scala
Exception in thread "main" java.lang.AssertionError: assertion failed: unresolved symbols: value <skolem>(line -1) when pickling Foo.scala
        at dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
        at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:725)
        at dotty.tools.dotc.transform.Pickler.run$$anonfun$9$$anonfun$8(Pickler.scala:63)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:87)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:87)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:318)
        at scala.collection.immutable.List.map(List.scala:246)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:319)
        at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:91)
        at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:165)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
        at dotty.tools.dotc.Run.runPhases$5(Run.scala:175)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:183)
        at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:190)
        at dotty.tools.dotc.Run.compileSources(Run.scala:127)
        at dotty.tools.dotc.Run.compile(Run.scala:110)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:38)
        at dotty.tools.dotc.Driver.process(Driver.scala:194)
        at dotty.tools.dotc.Driver.process(Driver.scala:163)
        at dotty.tools.dotc.Driver.process(Driver.scala:175)
        at dotty.tools.dotc.Driver.main(Driver.scala:202)
        at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

**Comment by nicolasstucki** [comment]
Not sure what is the correct thing to do here with the skolem type

**Comment by odersky** [comment]
This was fixed by c0a01a5
